shipa/provider.go

```diff
@@ -52,15 +64,26 @@ func Provider() *schema.Provider {
 }
 
 func providerConfigure(ctx context.Context, d *schema.ResourceData) (interface{}, diag.Diagnostics) {
+	var c *client.Client
+	var err error
+
 	token := d.Get("token").(string)
 	host := d.Get("host").(string)
+	adminEmail := d.Get("admin_email").(string)
+	adminPassword := d.Get("admin_password").(string)
 
 	// Warning or errors can be collected in a slice type
 	var diags diag.Diagnostics
 
-	c, err := client.NewClient(
-		client.WithHost(host),
-		client.WithToken(token))
+	if adminEmail != "" {
+		c, err = client.NewClient(
+			client.WithHost(host),
+			client.WithAdminAuth(adminEmail, adminPassword))
+	} else {
+		c, err = client.NewClient(
+			client.WithHost(host),
+			client.WithToken(token))
+	}
```

Comment:
i would maybe avoid this if/else

```
c, err = client.NewClient(
			client.WithHost(host),
			client.WithAdminAuth(adminEmail, adminPassword)
                        client.WithToken(token))
```

we can move validation of admnEmail and adminPassword to authenticate, and we trigger authenticate if token is empty, as you do already

Reply:
Currently we validate the token existence up front, whereas authentication check would run for each usage. Delaying the check that email and password are provided also could mean ignoring an empty token up front. What about a "WithAuth" that takes token, email, and password that gives an error of token is empty and both admin email and password are empty? I do still like the check in the authenticate method as well.